doc/sphinx/practical-tools/utilities.rst

```diff
@@ -7,32 +7,294 @@
+`Navigation/Start` in the menu or Ctrl-Home) for any buffer
+that uses the compiled files.  (Furthermore, you'll need to save the changes in your
+buffers to the file system before compiling them, such as with `File/Save all`
+on the menu.)
+
+If you want to build an entire multi-package project at once, you're on your own.
+There's currently no tooling to identify the internal dependencies between the packages
+(and thus the order in which to build the packages).
+
+
+Logical paths and the load path
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. TODO PR: add glossary entries for logical path and the load path
+   see modules.rst
+
+Coq commands such as :cmd:`Require` identify files with logical paths rather
+than file system paths so that scripts don't have to be modified to run on
+different computers.  The :cmd:`Print LoadPath` command displays the load path,
+which is a list of (logical path, physical path) pairs for directories.
+
+For example, you may see entries such as::
+
+  Logical Path / Physical path:
+  Bignums /home/jef/coq/lib/coq/user-contrib/Bignums
+  Bignums.BigZ /home/jef/coq/lib/coq/user-contrib/Bignums/BigZ
+  Ltac2 /home/jef/coq/lib/coq/user-contrib/Ltac2
+  Coq /home/jef/coq/lib/coq/theories
+  Coq.Numbers /home/jef/coq/lib/coq/theories/Numbers
+  Coq.Numbers.Natural /home/jef/coq/lib/coq/theories/Numbers/Natural
+  Coq.Numbers.Natural.Binary /home/jef/coq/lib/coq/theories/Numbers/Natural/Binary
+  Coq.Numbers.Integer /home/jef/coq/lib/coq/theories/Numbers/Integer
+  Coq.Arith /home/jef/coq/lib/coq/theories/Arith
+
+The components of each pair share suffixes, e.g. "Bignums.BigZ" and "Bignums/BigZ" or
+"Coq.Numbers.Natural" and "Numbers/Natural".  Physical pathnames should
+always use `/` rather than `\\`, even when running in on Windows.
+Packages with a physical path containing `user-contrib` were installed
+with the Coq binaries (e.g. `Ltac2`), with the Coq Platform or opam (e.g. `Bignums`)
+or perhaps by other means.  Note that, for these entries, the entire logical path
+appears in the directory name.
+Packages that begin with `Coq` were installed with the Coq binaries.  Note
+that the logical name `Coq` doesn't appear in the directory name.
+
+Each directory may contain multiple `.v`/`.vo` files.  For example,
+:n:`Require Import Coq.Numbers.Natural.Binary.NBinary` loads the file
+:n:`NBinary.vo` from the associated directory.  Note that a short name
+is often sufficient in :cmd:`Require` (in place of the fully-qualified
+name).
+
+The mechanisms for creating logical paths for your multi-file project are:
+
+- Through command line options passed on the command line to `coqc` and `coqide`,
+  principally :n:`-R <filesystem path> <logical name>`.  The resulting logical path
+  is the logical name followed by a suffix of the filesystem path (as for `Coq`
+  above).  Use these options directly if you must.  Command line options are described
+  :ref:`here <command-line-options>`.
```

Comment:
There is a subtle difference in preference with `-R` and `-Q` options.

When working on your own Coq project, I would say it is advisable to use `-R`. So if I had a `theories/` directory in my project, I would have the option `-R theories/ MyProject`.

When referencing another project in some directory that I want to use in my project I would prefer `-Q`. So if `OtherProject` lived in `path/to/other/` I would pass `-Q path/to/other/ OtherProject` to depend on it in MyProject.

The reason this is preferred is to not pollute the namespace. For example, OtherProject could have many files, many of which would be completely unrelated to MyProject. If I were to pass it as `-R` it would carry around a lot of needless baggage.

It might seem surprising that `-Q` is even enough here however. For example, if OtherProject had a deep folder structure, you might think that we need to specify all those subfolders too when depending on it. This is where some subtle Coq Require semantics step in. When you do `From OtherProject Require Import blah.`, Coq is actually able to search in subdirectories even if they were not specified! This means that coqdep (which is the workhorse finding dependencies) knows exactly where to find things.

This is the thing that Hugo worked on recently, which was to pin down precisely the semantics of this directory search when using `From`.

And finally, for the question of "Why have `-R` at all then?", the answer is that many Coq users do not like having to write `From MyProject Require Import myblah.` whilst inside MyProject, which is understandable.

Reply:
An overview of configuration won't describe every possible way to configure Coq.  We probably have more overlapping configuration knobs than we really need.  If `-Q` is useful for users I will mention it, but consider the following:

In the long term, it would be simpler if -Q -R and -I were not necessary, instead identifying packages solely by a directory name (including all subdirectories) and taking the full logical pathname from directory names.  That would be simpler for users.  So I wouldn't encourage the use of -Q consistent with this idea.

> When referencing another project in some directory that I want to use in my project I would prefer `-Q`. So if `OtherProject` lived in `path/to/other/` I would pass `-Q path/to/other/ OtherProject` to depend on it in MyProject.

That makes the _CoqProject system dependent: `path/to/other` will be not be in the directory containing _CoqProject or a subdirectory.  It may be an absolute path.  I thought that was not best practice, though it may be an easy shortcut that won't scale easily if you have multiple packages in your project (packages you maintain)--you likely would need to add the -Q entries into multiple `_CoqProject` files.

> The reason this is preferred is to not pollute the namespace.

"namespace" meaning the loadpath (logical names and directories) and when searching for files with `Require`.  I don't see how having more things in the loadpath is much of an issue, though `Print LoadPath` should probably have an option to show only the logical name.

As for the searching, I've been experimenting with creating my own `Bool.v`.  I can load my own IIRC and there's no indication that the builtin one has been hidden.  Users should have a way to report this.  The expected behavior is not documented.  This seems like a significant usability issue.

> This is the thing that Hugo worked on recently, which was to pin down precisely the semantics of this directory search when using `From`.

I remember that, but I've not been able trigger the "Required library qualid matches several files in path (found file1.vo, file2.vo, ...)." message.

> When you do `From OtherProject Require Import blah.`, Coq is actually able to search in subdirectories even if they were not specified!

Could you give a concrete example?

> And finally, for the question of "Why have `-R` at all then?", the answer is that many Coq users do not like having to write `From MyProject Require Import myblah.` whilst inside MyProject, which is understandable.

I think this reflects a weakness of how we currently resolve references.  Each package should be able to have local internal references without having to specify `From MyProject`, but reference resolution always uses the entire loadpath, i.e. global.  It's not so modular.  On the other hand, users probably don't want to have to write `From Coq Require Import Nat`, either.  I found 5800 uses of `Require` in the source tree but only 473 of them use `From`.

I'd like focus on adding useful info to the doc using what we have now.  I've found perhaps a dozen things that we could improve in configuration, but let's not get too diverting in analyzing those in depth.  I think the doc needs to clearly explain how conflicting/ambiguous references are handled.

Let me know your thoughts.  (The examples are invaluable, too.)

Reply:
>In the long term, it would be simpler if -Q -R and -I were not necessary, instead identifying packages solely by a directory name (including all subdirectories) and taking the full logical pathname from directory names. That would be simpler for users. So I wouldn't encourage the use of -Q consistent with this idea.

I believe that it is simpler to only think of -Q and -I when it comes to Coq projects, since internally this is how it is handled. `-R` is simply a shortcut to recursively do -Q for each subfolder if I am not mistaken.

>That makes the _CoqProject system dependent: path/to/other will be not be in the directory containing _CoqProject or a subdirectory. It may be an absolute path. I thought that was not best practice, though it may be an easy shortcut that won't scale easily if you have multiple packages in your project (packages you maintain)--you likely would need to add the -Q entries into multiple _CoqProject files.

`From ... Require Import ...` should be able to find external libraries (as long as lib and bin have been setup correctly) without the need to pass -Q. Typically the only reason to pass -Q are for folders in the very project you are working on, or for referencing projects which are not installed globally. In both cases, a relative path is typically easy to come by. Absolute paths for global install directories should not need to be given, since Coq should be able to find those itself.

>"namespace" meaning the loadpath (logical names and directories) and when searching for files with Require. I don't see how having more things in the loadpath is much of an issue, though Print LoadPath should probably have an option to show only the logical name.

If you are using `From ... Require Import ...` syntax, then the loadpath having many entries should typically not pose much of an issue. I believe that if you are using `Require Import ...` then Coq will have to start searching through all the loadpaths (consistently between coqc and coqdep at least). Your later example of Bool.v could demonstrate this problem. If I had many projects with a Bool.v in my loadpaths, how would I know which one was imported by `Require Import Bool.`. The answer I believe is, which ever came first in the list of load paths, but that is not something the user should be expected to remember.

>As for the searching, I've been experimenting with creating my own Bool.v. I can load my own IIRC and there's no indication that the builtin one has been hidden. Users should have a way to report this. The expected behavior is not documented. This seems like a significant usability issue.

It depends on how you require/import Bool. Inside your project, you should do `Require Import Bool` and expect your local version to be loaded. But if for whatever reason you need the stdlib version, then you can do `From Coq Require Import Bool`. That is not ambiguous for Coq to find.

>> When you do From OtherProject Require Import blah., Coq is actually able to search in subdirectories even if they were not specified!

>Could you give a concrete example?

For instance, if I installed [coq-equations](https://github.com/mattam82/Coq-Equations) globally using its `make install` or using opam (`opam install coq-equations`), then I should be able to open a new .v file and do `From Equations Require Import Prop.Classes`. At no point did I have to specify where coq-equations was actually installed, that's what I mean by `From` being able to search in directories not specified.

Another example that doesn't require an external package is if I had a directory structure like:
```
.
└── A
    ├── A.v
    └── B
        └── B.v
```

And I had only `-Q A A` in my _CoqProject, then if I did `From A Require Import B`, Coq would be able to find B.v. This is what I meant by not having to specify subdirectories. Of course, `Require Import A.B` would not work, because it is not in the loadpath. And `-R A A` would have done something equivalent to `-Q A A`, `-Q A/B A.V` in the background.

Reply:
>Of course, Require Import A.B would not work, because it is not in the loadpath

I don't know what you mean by "not in the loadpath", it doesn't work because its name is A.B.B not A.B and Require doesn't insert components in the middle of qualids

Reply:
Right my mistake, I meant `Require Import A.B.B` would not work. The module for the file A/B/B.v would not be seen by Coq correct?

Reply:
It would work. -Q and -R are the same except Require without From with -Q needs more explicit qualids.

Reply:
OK I stand corrected. At least it's good that we are attempting to write this documentation. It still confuses me. 0:-)